utils/s2n_atomic.h

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#pragma once
+
+#include <signal.h>
+
+#include "utils/s2n_result.h"
+
+/* s2n-tls allows s2n_send and s2n_recv to be called concurrently.
+ * There are a handful of values that both methods need to access.
+ * However, C99 has no concept of concurrency, so provides no atomic data types.
+ *
+ * Traditionally, s2n-tls uses sig_atomic_t for its weak guarantee of atomicity for interrupts.
+ */
+typedef sig_atomic_t s2n_atomic;
+
+/* These methods use compiler atomic built-ins if available and lock-free, but otherwise
+ * rely on setting / clearing a small value generally being atomic in practice.
+ */
+S2N_RESULT s2n_atomic_init();
+void s2n_atomic_set(s2n_atomic *var);
```

Comment:
It's probably worth copying the naming conventions from [rust](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html) or [cpp](https://en.cppreference.com/w/cpp/atomic/atomic). That would be `store` instead of `set`, `load` instead of `check`.

Reply:
Idk, `store` implies that you can specify a value, instead of it always storing true.
And how about "test" instead of "check"? That'd be inline with the builtin [__atomic_test_and_set](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html#index-_005f_005fatomic_005ftest_005fand_005fset)

Reply:
After our offline conversation, I replaced set/clear with load. But I've revert that-- I think set/clear is safer. See the [KeyUpdate bug](https://github.com/aws/s2n-tls/blob/main/tls/s2n_key_update.c#L39) I use as an example in the description: because that code blindly sets `conn->key_update_pending` to `key_update_request`, it could be setting key_update_pending back to 0. But only the writer is allowed to set key_update_pending back to 0, or else we lose key updates. Developers need to be very clear on whether they're setting or clearing the flag.